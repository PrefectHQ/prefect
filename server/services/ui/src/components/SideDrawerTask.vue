<script>
export default {
  filters: {
    typeClass: val => val.split('.').pop()
  },
  props: {
    genericInput: {
      required: true,
      type: Object,
      validator: input => {
        return !!input.taskId
      }
    }
  },
  data() {
    return {
      task: null
    }
  },
  apollo: {
    task: {
      query: require('@/graphql/Task/task-drawer.gql'),
      variables() {
        return { id: this.genericInput.taskId }
      },
      pollInterval: 1000,
      update: data => data.task_by_pk
    }
  }
}
</script>

<template>
  <v-container class="side-drawer-flow-runs mb-10 pb-10">
    <v-layout
      v-if="$apollo.loading || !task"
      align-center
      column
      justify-center
    >
      <v-progress-circular indeterminate color="primary" />
    </v-layout>
    <v-layout v-else column>
      <v-flex xs12>
        <div class="title">
          <router-link
            class="link"
            :to="{
              name: 'task',
              params: { id: task.id }
            }"
          >
            {{ task.name }}
          </router-link>
        </div>
        <div v-if="task.description" class="subtitle-1">
          {{ task.description }}
        </div>
      </v-flex>
      <v-flex xs12 class="mt-4">
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Flow
          </v-flex>
          <v-flex xs6 class="text-right body-1">
            <router-link
              class="link"
              :to="{ name: 'flow', params: { id: task.flow.id } }"
            >
              {{ task.flow.name }}
            </router-link>
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Latest Run State
          </v-flex>
          <v-flex xs6 class="text-right body-1">
            <span v-if="task.task_runs.length" class="pr-1">
              <v-icon small :color="task.task_runs[0].state">
                brightness_1
              </v-icon>
            </span>
            {{ task.task_runs.length ? task.task_runs[0].state : 'No Runs' }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Mapped
          </v-flex>
          <v-flex xs6 class="text-right body-1">
            {{ task.mapped ? 'Yes' : 'No' }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Slug
          </v-flex>
          <v-flex id="slug" xs6 class="text-right body-1 word-wrap">
            {{ task.slug }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Max Retries
          </v-flex>
          <v-flex xs6 class="text-right body-1">
            {{ task.max_retries | number }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Retry Delay
          </v-flex>
          <v-flex xs6 class="text-right body-1">
            {{ task.retry_delay | duration }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Trigger
          </v-flex>
          <v-flex xs6 class="text-right body-1 word-wrap">
            {{ task.trigger }}
          </v-flex>
        </v-layout>
        <v-layout class="my-2">
          <v-flex xs6 class="text-left body-1 font-weight-medium">
            Class
          </v-flex>
          <v-flex xs6 class="text-right body-1 word-wrap">
            {{ task.type | typeClass }}
          </v-flex>
        </v-layout>
      </v-flex>
    </v-layout>
  </v-container>
</template>

<style lang="scss" scoped>
.side-drawer-flow-runs {
  overflow-y: scroll;
}

.word-wrap {
  word-wrap: break-word;
}
</style>
