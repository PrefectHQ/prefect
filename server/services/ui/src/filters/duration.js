import { durationToEnglish } from '@/utils/dateTime'

export default function(duration) {
  return durationToEnglish(duration)
}
