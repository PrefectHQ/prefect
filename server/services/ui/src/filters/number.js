import numeral from 'numeral'

export default function(number) {
  return numeral(number).format('0,0')
}
