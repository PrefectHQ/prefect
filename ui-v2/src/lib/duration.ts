import {
	formatDuration,
	intervalToDuration,
	type Duration,
	formatDistanceStrict,
} from "date-fns";

export function secondsToString(input: number): string {
	const duration: Duration = intervalToDuration({
		start: 0,
		end: input * 1000, // convert seconds to milliseconds
	});

	return formatDuration(duration, {
		zero: false,
		delimiter: " ",
	}).trim();
}

export function secondsToApproximateString(input: number): string {
	return formatDistanceStrict(0, input * 1000, {
		addSuffix: false,
	});
}
