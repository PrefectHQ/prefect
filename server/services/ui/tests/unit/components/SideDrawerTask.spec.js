import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Vuex from 'vuex'
import SideDrawerTask from '@/components/SideDrawerTask'
import duration from '@/filters/duration'
import number from '@/filters/number'

jest.mock('@/graphql/Task/task-drawer.gql', () => {
  return 'a graphql string'
})

const localVue = createLocalVue()
localVue.use(Vuex)

describe('SideDrawerTask', () => {
  const wrapper = shallowMount(SideDrawerTask, {
    data() {
      return {
        task: {
          id: 'ed8527c7-f810-4b1c-a400-afaebc0ce244',
          name: 'Pull SYMC price',
          trigger: '<function all_successful at 0x7fbfdd288c20>',
          type: 'prefect.tasks.core.function.FunctionTask',
          slug: '8446ca22-7038-4ae0-b702-70b2fb0b5a17',
          task_runs: {},
          flow: {
            id: 'e1cc3664-2445-4d6c-86b4-4fdd885a9cba',
            name: 'Core v0.6.0: Stock Flow'
          }
        }
      }
    },
    propsData: {
      genericInput: {
        taskId: 'b80d9a19-8339-44a4-a695-3231cbb72ae6'
      }
    },
    mocks: {
      $apollo: {
        loading: false,
        data: {}
      }
    },
    stubs: {
      RouterLink: RouterLinkStub,
      PrefectSchedule: true,
      VBtn: true,
      VContainer: true,
      VLayout: true,
      VFlex: true
    },
    localVue,
    filters: {
      number,
      duration
    }
  })
  it('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })
  it('shows the task slug', () => {
    expect(wrapper.find('#slug').text()).toBe(
      '8446ca22-7038-4ae0-b702-70b2fb0b5a17'
    )
  })
  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
