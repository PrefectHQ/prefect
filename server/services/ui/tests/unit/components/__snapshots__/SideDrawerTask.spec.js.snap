// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideDrawerTask matches the snapshot 1`] = `
<vcontainer-stub
  class="side-drawer-flow-runs mb-10 pb-10"
>
  <vlayout-stub
    column=""
  >
    <vflex-stub
      xs12=""
    >
      <div
        class="title"
      >
        <a
          class="link"
        >
          
          Pull SYMC price
        
        </a>
      </div>
       
      <!---->
    </vflex-stub>
     
    <vflex-stub
      class="mt-4"
      xs12=""
    >
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Flow
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1"
          xs6=""
        >
          <a
            class="link"
          >
            
            Core v0.6.0: Stock Flow
          
          </a>
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Latest Run State
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1"
          xs6=""
        >
          <!---->
          
          No Runs
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Mapped
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1"
          xs6=""
        >
          
          No
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Slug
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1 word-wrap"
          id="slug"
          xs6=""
        >
          
          8446ca22-7038-4ae0-b702-70b2fb0b5a17
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Max Retries
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1"
          xs6=""
        >
          
          0
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Retry Delay
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1"
          xs6=""
        >
          
          
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Trigger
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1 word-wrap"
          xs6=""
        >
          
          &lt;function all_successful at 0x7fbfdd288c20&gt;
        
        </vflex-stub>
      </vlayout-stub>
       
      <vlayout-stub
        class="my-2"
      >
        <vflex-stub
          class="text-left body-1 font-weight-medium"
          xs6=""
        >
          
          Class
        
        </vflex-stub>
         
        <vflex-stub
          class="text-right body-1 word-wrap"
          xs6=""
        >
          
          FunctionTask
        
        </vflex-stub>
      </vlayout-stub>
    </vflex-stub>
  </vlayout-stub>
</vcontainer-stub>
`;
